devtools/x/src/installer.rs

```diff
@@ -1,42 +1,107 @@
-            .arg(name)
-            .arg("--version")
-            .arg(version)
-            .run();
-        if installers.is_err() {
+impl Installer {
+    // Given the name of a tool in the tools list, return the version.
+    fn cargo_installation(&self, name: &str) -> Option<&CargoInstallation> {
+        self.cargo_installations
+            .as_slice()
+            .iter()
+            .find_map(|(x, y)| if x.eq(&name) { Some(y) } else { None })
+    }
+
+    pub fn install_if_needed(&self, name: &str) -> bool {
+        match &self.cargo_installation(name) {
+            Some(cargo_installation) => {
+                install_if_needed(&self.cargo_config, name, &cargo_installation)
+            }
+            None => {
+                info!("No version of tool {} is specified ", name);
+                false
+            }
+        }
+    }
+
+    #[allow(dead_code)]
+    fn check(&self, name: &str) -> bool {
+        match &self.cargo_installation(name) {
+            Some(cargo_installation) => check_installed(name, &cargo_installation.version),
+            None => {
+                info!("No version of tool {} is specified ", name);
+                false
+            }
+        }
+    }
+
+    pub fn check_all(&self) -> bool {
+        let iter = self
+            .cargo_installations
+            .as_slice()
+            .iter()
+            .map(|(name, installation)| (name, &installation.version))
+            .collect::<Vec<(&String, &String)>>();
+        check_all(iter.as_slice())
+    }
+
+    pub fn install_all(&self) -> bool {
+        install_all(&self.cargo_config, &self.cargo_installations.as_slice())
+    }
+}
+
+pub fn install_if_needed(
+    cargo_config: &CargoConfig,
+    name: &str,
+    installation: &CargoInstallation,
+) -> bool {
+    if !check_installed(name, &installation.version) {
+        info!("Installing {} {}", name, installation.version);
+        //prevent recursive install attempts of sccache.
+        let mut cmd = Cargo::new(&cargo_config, "install", false);
+        if let Some(features) = &installation.features {
+            if !features.is_empty() {
+                cmd.arg("--features");
+                cmd.args(features);
+            }
+        }
+        if let Some(git_url) = &installation.git {
+            cmd.arg("--git");
+            cmd.arg(git_url);
+            if let Some(git_rev) = &installation.git_rev {
+                cmd.arg("--rev");
+                cmd.arg(git_rev);
+            }
+        } else {
+            cmd.arg("--version").arg(&installation.version);
+        }
+        cmd.arg(name);
+
+        let result = cmd.run();
+        if result.is_err() {
             info!(
-                "Could not install {} {}, check x.toml to ensure tool exists and is not yanked.",
-                name, version
+                "Could not install {} {}, check x.toml to ensure tool exists and is not yanked, or provide a git revision.",
```

Comment:
How would a git revision be provided? Could the message be clearer?

Reply:
updated